### ISSUE TITLE
cv2.remap in the cv2.INTER_LINEAR has some problem in border values.

### ISSUE THREAD

**Issue by TaooCAI** [opencv/opencv#21148]
##### System information (version)
- opencv-python => 4.5.2.54
- numpy => 1.19.5
- python => 3.6.13
- Operating System / Platform => Linux 64 Bit

##### Detailed description

When using cv2.remap in the cv2.INTER_LINEAR interpolation with border default value (0), if the mapping has a value in [h-1, h] or [w-1, w], h and w is the height and width of the mapping and the image for warping, the finally warped image has values in invalid regions. It seems that cv2.remap may be not careful with float numbers rounding ?

##### Steps to reproduce

```.py
import cv2
import numpy as np


def test(img, mapping):
    # remap
    warped_img = cv2.remap(img, mapping, None, interpolation=cv2.INTER_LINEAR)
    cv2.imwrite('warped_img.png', warped_img)
    # bug mask
    h, w, _ = mapping.shape
    mask = (mapping[:, :, 0] > (w-1)) & ((mapping[:, :, 0] < w)) | (mapping[:, :, 1] > (h-1)) & (mapping[:, :, 1] < h)
    cv2.imwrite('mask.png', mask*255)


if __name__ == "__main__":
    img = cv2.imread('img.png', cv2.IMREAD_COLOR)
    mapping = np.load('mapping.npz')['mapping']
    test(img, mapping)
```
img.png
![img](https://user-images.githubusercontent.com/59591644/143824501-04851b88-d324-4f7b-91e3-4230d1cb8c47.png)
warped_img.png
![warped_img](https://user-images.githubusercontent.com/59591644/143824525-d279a4a7-cdfa-4c6e-b208-3f2c023ce98b.png)
mask.png
![mask](https://user-images.githubusercontent.com/59591644/143824542-3fe2fcf1-a621-41c0-a8f1-ae5c2c23f2c8.png)

mapping.npz
[mapping.zip](https://github.com/opencv/opencv/files/7616264/mapping.zip)
##### Issue submission checklist

 - [x] I report the issue, it's not a question
   
 - [x] I checked the problem with documentation, FAQ, open issues,
       forum.opencv.org, Stack Overflow, etc and have not found solution
   
 - [x] I updated to latest OpenCV version and the issue is still there
   
 - [x] There is reproducer code and related data files: videos, images, onnx, etc

**Comment by AleksandrPanov** [comment]
@TaooCAI there are indexes with `x` in (w-1, w-0.5) and `y` in (h-1, h-0.5). `remap` with INTER_LINEAR (and other interpolations) use 'cvRound' for border indexes that rounds (h-1, h-0.5)  to `h-1` and rounds (w-1, w-0.5) to `w-1`.
After rounding, the indices are in the valid range and are displayed in the images as lines.

You can fix `mapping` by add `0.5` to indexes with `x` in (w-1, w-0.5) and `y` in (h-1, h-0.5):
```
mask_x = np.where(mapping[:, :, 0] > (w - 1), 0.5, 0.)
mask_y = np.where(mapping[:, :, 1] > (h - 1), 0.5, 0.)

mapping[:, :, 0] += mask_x
mapping[:, :, 1] += mask_y
```
![INTER_LINEAR](https://user-images.githubusercontent.com/22337800/144893318-9415a617-c45e-4995-bf18-521c10411ed4.png)

[repr.py.txt](https://github.com/opencv/opencv/files/7662449/repr.py.txt)

**Comment by TaooCAI** [comment]
> @TaooCAI there are indexes with `x` in (w-1, w-0.5) and `y` in (h-1, h-0.5). `remap` with INTER_LINEAR (and other interpolations) uses 'cvRound' for border indexes that rounds (h-1, h-0.5) to `h-1` and rounds (w-1, w-0.5) to `w-1`. After rounding, the indexes are in the valid range and are displayed in the images as lines.
> 
> You can fix `mapping` by adding `0.5` to indexes with `x` in (w-1, w-0.5) and `y` in (h-1, h-0.5):
> 
> ```
> mask_x = np.where(mapping[:, :, 0] > (w - 1), 0.5, 0.)
> mask_y = np.where(mapping[:, :, 1] > (h - 1), 0.5, 0.)
> 
> mapping[:, :, 0] += mask_x
> mapping[:, :, 1] += mask_y
> ```
> 
> ![INTER_LINEAR](https://user-images.githubusercontent.com/22337800/144893318-9415a617-c45e-4995-bf18-521c10411ed4.png)
> 
> [repr.py.txt](https://github.com/opencv/opencv/files/7662449/repr.py.txt)

This is indeed an effective solution, but it is better to solve such problem in `cv2.remap`. Otherwise, each time when we use the `remap`, we need to add an extra lines of code to preprocess, confuse others, explain repeatedly. When we occasionally forget to do that, some problems occur.
